napari/_vispy/vispy_image_layer.py

```diff
@@ -93,6 +93,102 @@ def _on_contrast_limits_change(self):
         else:
             self.node.clim = self.layer.contrast_limits
 
+    def _on_scale_change(self):
+        self.scale = [
+            self.layer.scale[d] * self.layer._scale_view[d]
+            for d in self.layer.dims.displayed[::-1]
+        ]
+        if self.layer.pyramid:
+            self.layer.top_left = self.find_top_left()
+        self.layer.position = self._transform_position(self._position)
+
+    def _on_translate_change(self):
+        self.translate = [
+            self.layer.translate[d] + self.layer._translate_view[d]
+            for d in self.layer.dims.displayed[::-1]
+        ]
+        self.layer.position = self._transform_position(self._position)
+
+    def compute_data_level(self, size):
+        """Computed what level of the pyramid should be viewed given the
+        current size of the requested field of view.
+
+        Parameters
+        ----------
+        size : 2-tuple
+            Requested size of field of view in image coordinates
+
+        Returns
+        ----------
+        level : int
+            Level of the pyramid to be viewing.
+        """
+        # Convert requested field of view from the camera into log units
+        size = np.log2(np.max(size))
+
+        # Max allowed tile in log units
+        max_size = np.log2(self.layer._max_tile_shape)
+
+        # Allow for more than 2x coverage of field of view with max tile
+        diff = size - max_size + 1.25
+
+        # Find closed downsample level to diff
+        ds = self.layer.level_downsamples[:, self.layer.dims.displayed].max(
+            axis=1
+        )
+        level = np.argmin(abs(np.log2(ds) - diff))
+
+        return level
+
+    def find_top_left(self):
+        """Finds the top left pixel of the canvas. Depends on the current
+        pan and zoom position
+
+        Returns
+        ----------
+        top_left : tuple of int
+            Coordinates of top left pixel.
+        """
+        nd = self.layer.dims.ndisplay
+        # Find image coordinate of top left canvas pixel
+        if self.node.canvas is not None:
+            transform = self.node.canvas.scene.node_transform(self.node)
+            pos = (
+                transform.map([0, 0])[:nd]
+                + self.translate[:nd] / self.scale[:nd]
+            )
+        else:
+            pos = [0] * nd
+
+        top_left = np.zeros(self.layer.ndim, dtype=int)
+        for i, d in enumerate(self.layer.dims.displayed[::-1]):
+            top_left[d] = pos[i]
+
+        # Clip according to the max image shape
+        top_left = np.clip(
+            top_left, 0, np.subtract(self.layer.level_shapes[0], 1)
+        )
+
+        # Convert to offset for image array
+        rounding_factor = self.layer._max_tile_shape / 4
+        top_left = rounding_factor * np.floor(top_left / rounding_factor)
+
+        return top_left.astype(int)
+
+    def on_draw(self, event):
```

Comment:
Could this be renamed `on_draw_pyramid` and only get connected when the image is a pyramid? It's a lot of extra calls when displaying a random image, which won't be super expensive but still feels wasteful, especially one of the main design goals of napari is to be blazing fast.

Reply:
Checking that if statement is really negligible I think. This only gets called when the image changes or the mouse moves, and will make it easier if a layer switches from being a pyramid to a non-pyramid, or visa versa. Otherwise we need to connect / disconnect. I can still change if there is some measurable performance difference, but I think it will be in the microseconds and not happening always.

Reply:
I worry that all the stuff happening during mouse moves is adding up and will continue to do so. When will a layer switch “live” rather than at creation time?